src/preprocessing/preprocess.ts

```diff
@@ -0,0 +1,25 @@
+import { Config } from '../config';
+import { modules as preprocessingModules } from './preprocessing-modules/index';
+
+/**
+ * Runs the preprocessing modules on the given AMP code.
+ *
+ * @param {string} amp AMP code to preprocess
+ * @param {!Config} config Global config
+ * @param {Array<string>=} modules Module whitelist (leave empty to run all)
+ * @return {!Promise<string>} Preprocessed AMP code
+ */
+export async function preprocessAMP(
+  amp: string,
+  config: Config,
+  modules?: string[]
+): Promise<string> {
+  const moduleSet = new Set(modules || []);
+  for (const module of preprocessingModules) {
```

Comment:
Can preprocessingModules be a name to module mapping of the available modules rather than an array. That way we can iterate though moduleSet instead and just fetch from map and process.

e.g. 

    for (const moduleName of modules) {
      if (preprocessingModules[moduleName]) {
        amp = await module.process(amp, config);
      }
    }

Reply:
I thought about that: however, the issue becomes order. It's important to e.g. run validation before injecting things into `<head>`.

If we iterate the user-provided array, then the user needs to be aware of these constraints and ensure the correct order. An alternative would be to build a more sophisticated dependency system (e.g. each module has a `dependsOn` or so), but that sounds like it'd add too much complexity.

Reply:
How about the client provide the modules they want to opt out of?  So that way, they need not concern themselves with order and are opt in to all this goodness by default.

Reply:
That's a really good idea. It makes the default of empty a lot more sane.

I also moved this to the global config, since it seems unnecessary to have it as a separate parameter. PTAL.